File: apps/desktop/src/features/chat/receipts.ts
/**
 * Read receipts, derived from read markers.
 *
 * The server stores one marker per person per channel - "read up to here" -
 * and never a row per message. Everything below is the arithmetic that turns
 * those markers back into "who has seen this message", which is why there is
 * no receipt table to keep in step with anything.
 *
 * Two questions, and they are not the same one:
 *
 * - `seenBy` - who has seen *this* message. Every marker at or past it. That
 *   is what the dialog lists.
 * - `anchorReceipts` - where each person's face is drawn. A reader is shown
 *   once, on the newest message of yours they have read, the way Messenger
 *   does it. Drawing everybody against every message would repeat the same
 *   four avatars down the whole conversation and say nothing new each time.
 */
import type { ChannelReadReceipt } from '@betweenus/shared-types';

/** How many faces the row draws before it starts counting instead. */
export const SEEN_BY_FACES = 4;

/** The shape both callers need from a message; the real one carries far more. */
export interface ReceiptMessage {
  id: string;
  createdAt: string;
  authorId: string;
}

const at = (iso: string): number => new Date(iso).getTime();

/** Everyone whose marker is at or past this message. Newest reader first. */
export function seenBy(
  message: Pick<ReceiptMessage, 'createdAt'>,
  receipts: ChannelReadReceipt[],
): ChannelReadReceipt[] {
  const sent = at(message.createdAt);
  return receipts
    .filter((receipt) => at(receipt.readAt) >= sent)
    .sort((a, b) => at(a.readAt) - at(b.readAt));
}

/**
 * messageId -> the readers whose row it is, for your own messages only.
 *
 * Somebody else's message never carries a receipt: it is their message that
 * was read, and telling you who else has read it is a different feature.
 */
export function anchorReceipts(
  messages: ReceiptMessage[],
  receipts: ChannelReadReceipt[],
  meId: string | undefined,
): Record<string, ChannelReadReceipt[]> {
  if (!meId) return {};
  const mine = messages
    .filter((message) => message.authorId === meId)
    .sort((a, b) => at(a.createdAt) - at(b.createdAt));
  if (mine.length === 0) return {};

  const anchors: Record<string, ChannelReadReceipt[]> = {};
  for (const receipt of receipts) {
    const read = at(receipt.readAt);
    // The newest of your messages this marker has reached. A marker older than
    // everything on screen anchors nowhere, which is the "has not seen it yet"
    // case and draws nothing at all.
    let newest: ReceiptMessage | null = null;
    for (const message of mine) {
      if (at(message.createdAt) <= read) newest = message;
      else break;
    }
    if (!newest) continue;
    (anchors[newest.id] ??= []).push(receipt);
  }

  // Oldest reader first within a row, so the faces stop shuffling every time
  // somebody else opens the channel.
  for (const row of Object.values(anchors)) row.sort((a, b) => at(a.readAt) - at(b.readAt));
  return anchors;
}

/** "Seen by Ana", "Seen by Ana and Bo", "Seen by Ana, Bo and 3 others". */
export function seenByLabel(receipts: ChannelReadReceipt[]): string {
  const names = receipts.map((receipt) => receipt.user.displayName || receipt.user.username);
  if (names.length === 0) return 'Not seen yet';
  if (names.length === 1) return `Seen by ${names[0]}`;
  if (names.length === 2) return `Seen by ${names[0]} and ${names[1]}`;
  const rest = names.length - 2;
  return `Seen by ${names[0]}, ${names[1]} and ${rest} other${rest === 1 ? '' : 's'}`;
}

File: apps/android/core/src/main/java/com/aatech/betweenus/core/store/Receipts.kt
package com.aatech.betweenus.core.store

import com.aatech.betweenus.core.data.ChannelReadReceipt
import java.time.Instant

/**
 * Read receipts, derived from read markers.
 *
 * The port of `apps/desktop/src/features/chat/receipts.ts`, and the same
 * arithmetic for the same reason: the server stores one marker per person per
 * channel - "read up to here" - and never a row per message, so "who has seen
 * this message" is "whose marker is at or past its timestamp".
 *
 * Two questions, and they are not the same one:
 *
 * - [seenBy] - who has seen *this* message. What the dialog lists.
 * - [anchorReceipts] - where each person's face is drawn. Once, on the newest
 *   message of yours they have read. Drawing everybody against every message
 *   would repeat the same four faces down the whole conversation.
 */
object Receipts {
    /** How many faces the row draws before it starts counting instead. */
    const val FACES = 4

    private fun at(iso: String): Long =
        runCatching { Instant.parse(iso).toEpochMilli() }.getOrDefault(0L)

    /** Everyone whose marker is at or past this message. Oldest reader first. */
    fun seenBy(sentAt: String, receipts: List<ChannelReadReceipt>): List<ChannelReadReceipt> {
        val sent = at(sentAt)
        return receipts.filter { at(it.readAt) >= sent }.sortedBy { at(it.readAt) }
    }

    /**
     * messageId -> the readers whose row it is, for [selfId]'s messages only.
     *
     * Somebody else's message never carries a receipt: it is their message that
     * was read, and telling you who else has read it is a different feature.
     */
    fun anchorReceipts(
        messages: List<ReadableMessage>,
        receipts: List<ChannelReadReceipt>,
        selfId: String?,
    ): Map<String, List<ChannelReadReceipt>> {
        if (selfId == null) return emptyMap()
        val mine = messages
            .filter { it.message.author.id == selfId }
            .sortedBy { at(it.message.createdAt) }
        if (mine.isEmpty()) return emptyMap()

        val anchors = mutableMapOf<String, MutableList<ChannelReadReceipt>>()
        for (receipt in receipts) {
            val read = at(receipt.readAt)
            // The newest message this marker has reached. A marker older than
            // everything on screen anchors nowhere, which is the "has not seen
            // it yet" case and draws nothing at all.
            val newest = mine.lastOrNull { at(it.message.createdAt) <= read } ?: continue
            anchors.getOrPut(newest.id) { mutableListOf() }.add(receipt)
        }

        // Oldest reader first within a row, so the faces stop shuffling every
        // time somebody else opens the channel.
        return anchors.mapValues { (_, row) -> row.sortedBy { at(it.readAt) } }
    }

    /** "Seen by Ana", "Seen by Ana and Bo", "Seen by Ana, Bo and 3 others". */
    fun seenByLabel(receipts: List<ChannelReadReceipt>): String {
        val names = receipts.map { it.user.label }
        return when {
            names.isEmpty() -> "Not seen yet"
            names.size == 1 -> "Seen by ${names[0]}"
            names.size == 2 -> "Seen by ${names[0]} and ${names[1]}"
            else -> {
                val rest = names.size - 2
                "Seen by ${names[0]}, ${names[1]} and $rest other" + if (rest == 1) "" else "s"
            }
        }
    }
}
